### PR TITLE
Add an 'Exclude' option to 'Filter widget type'

### LINKED ISSUE #30764: Add "Exclude ..." option to Filter widget type dropdown
There’s no option to create an “exclude” type field filter in a native query template tag. You have to select “Date Filter”, and then the full menu of options comes up when selecting the parameter value, including the “Exclude …” option. 

image

The "Exclude..." parameter type is available if the "Date Filter" option is selected:
image

[Slack context](https://metaboat.slack.com/archives/C057T1QTB3L/p1684162994825709): Bruno: "Pretty sure we overlooked it"

### PR DESCRIPTION
**Pull request by kienstra**
Closes https://github.com/metabase/metabase/issues/30764

### Description

Adds an 'Exclude' option to 'Filter widget type':

<div float="left">
Screenshot 2024-01-17 at 10 30 56 AM

</div>

…so native queries can exclude by date, without having to select 'Date Filter' -> 'Exclude'

### How to verify

1. Create a new question
2. Click 'Sample Database'
3. Click the 'Accounts' table
4. Click 'View the SQL' 
5. Click 'Convert this question to SQL':

Screenshot 2024-01-17 at 9 35 43 AM

6. Replace the SQL with this:

```sql
SELECT *
FROM accounts
WHERE {{dateOfCreation}}
```

7. From 'Variable type',  select 'Field Filter':
8. From 'Field to map to', select 'Accounts' -> 'Created At' (or any other date field)
9. From 'Filter widget type', select 'Exclude' (the new option in this PR):

Screenshot 2024-01-17 at 9 40 46 AM

10. Exclude September
11. Click 'Run query'
12. Expected: September doesn't show in `'CREATED_AT'` anymore:

https://github.com/kienstra/metabase/assets/4063887/3a4b7ba1-d968-443f-840a-9f7aae0d9473

(No audio in either video)

### Demo

https://github.com/metabase/metabase/assets/4063887/d5b19f0a-43fc-41e0-a883-c3cf28da7da9

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/components/DateExcludeWidget/DateExcludeWidget.tsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  const initial = ["!=", ["field", null, {}]];
+  const [filter, setFilter] = useState(
+    dateParameterValueToMBQL(value, null) || initial,
+  );
+
+  const filterSelected = EXCLUDE_OPERATORS.some(({ test }) => test(filter));
+  const commitAndClose = (newFilter?: any) => {
+    setValue(filterToUrlEncoded(newFilter || filter));
+    onClose?.();
+  };
```
Comment (kienstra) [inline frontend/src/metabase/components/DateExcludeWidget/DateExcludeWidget.tsx:36]: Copy-pasted from [DateAllOptionsWidget.tsx](https://github.com/metabase/metabase/blob/c42aa270842c2292f87fa9d17bd8b27e9a1087db/frontend/src/metabase/components/DateAllOptionsWidget/DateAllOptionsWidget.tsx#L35-L38)

File: frontend/src/metabase/components/DateExcludeWidget/DateExcludeWidget.styled.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+export const Container = styled.div`
+  display: flex;
+  flex-wrap: no-wrap;
+  justify-content: space-between;
+  align-items: center;
+  border-top: 1px solid ${color("border")};
+  border-bottom: 1px solid ${color("border")};
+  padding: ${space(2)};
+`;
```
Comment (kienstra) [inline frontend/src/metabase/components/DateExcludeWidget/DateExcludeWidget.styled.tsx:13]: Copy-pasted from [DatePickerFooter.styled.tsx](https://github.com/metabase/metabase/blob/519e2d327949de1de6b028bbaa22833806bf94c0/frontend/src/metabase/admin/datamodel/components/filters/pickers/DatePicker/DatePickerFooter.styled.tsx#L7-L14), but with a `border-bottom`, and the same `padding` on all sides:

Screenshot 2024-01-17 at 11 02 42 AM

**Comment by darksciencebase** [comment]
cc @NevRA

**Comment by NevRA** [comment]
Thank you for your contribution. But now we are in the process of migrating this functionality to CLJS and are not making changes to metabase-lib. Therefore I have to close this PR. We will think about how to make the collaboration process more transparent.

**Comment by kienstra** [comment]
Hi @NevRA,
All good, thanks for letting me know! Makes sense to close this PR.

**Comment by kienstra** [comment]
>To prevent spending significant effort and time in vain next time, maybe just quickly check with the devs before you start working on an issue.

Thanks, that's a great idea! Makes sense that fixes are safer than features.

**Pull request closed without merging**